### PR TITLE
feat: classes of C⋆-algebras

### PR DESCRIPTION
**Pull request by j-loreaux**
This adds four classes of C⋆-algebras (unital / non-unital, commutative / noncommutative). We only allow for complex scalars, as the theory for real scalars, perhaps counterintuitively, depends on the theory for complex scalars.

For all existing `CStarRing` instances, we also include the appropriate `CStarAlgebra` instance (so long as it actually makes sense). The purpose of this is (a) to reduce typing and the size of `variable` statements, but also (b) to hopefully reduce burden on type class synthesis.

Note: we leave `PartialOrder` and `StarOrderedRing` unbundled from `CStarAlgebra`. This is because, even in the most common cases (e.g., `ℂ`), the partial order can be controversial as a global instance. Moreover, including it wouldn't match the literature (although this point is not in and of itself sufficient justification). So, if your statement about C⋆-algebras requires `≤`, it is necessary to add these classes.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/ContinuousFunction/Bounded.lean [hunk 1]
```diff
@@ -1245,6 +1263,23 @@ instance instNormedCommRing [NormedCommRing R] : NormedCommRing (α →ᵇ R) wh
...
+
+instance [IsScalarTower 𝕜 β β] : IsScalarTower 𝕜 (α →ᵇ β) (α →ᵇ β) where
+  smul_assoc _ _ _ := ext fun _ ↦ smul_mul_assoc ..
+
+instance [SMulCommClass 𝕜 β β] : SMulCommClass 𝕜 (α →ᵇ β) (α →ᵇ β) where
+  smul_comm _ _ _ := ext fun _ ↦ (mul_smul_comm ..).symm
+
+example : instPseudoMetricSpace (α := α) (β := β) = NonUnitalSeminormedRing.toPseudoMetricSpace :=
+  rfl
+
```
Comment (j-loreaux) [inline Mathlib/Topology/ContinuousFunction/Bounded.lean:1280]:
```suggestion
```

File: Mathlib/Analysis/CStarAlgebra/Classes.lean [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+-/
+
+/-- The class of non-unital (complex) C⋆-algebras. -/
+class NonUnitalCStarAlgebra (A : Type*) extends NonUnitalNormedRing A, StarRing A, CompleteSpace A,
+    CStarRing A, NormedSpace ℂ A, IsScalarTower ℂ A A, SMulCommClass ℂ A A, StarModule ℂ A where
+
+/-- The class of non-unital commutative (complex) C⋆-algebras. -/
+class NonUnitalCommCStarAlgebra (A : Type*) extends NonUnitalCStarAlgebra A where
+  mul_comm : ∀ (a b : A), a * b = b * a
```
Comment (eric-wieser) [inline Mathlib/Analysis/CStarAlgebra/Classes.lean:29]: I would suggest
```suggestion
class NonUnitalCommCStarAlgebra (A : Type*) 
  extends NonUnitalNormedCommRing A, NonUnitalCStarAlgebra A
```

to save having to write the extra instances.

File: Mathlib/Analysis/CStarAlgebra/Classes.lean [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+instance (priority := 100) CommCStarAlgebra.toNormedCommRing
+    (A : Type*) [CommCStarAlgebra A] : NormedCommRing A where
+  mul_comm := mul_comm
+
+instance (priority := 100) NonUnitalCommCStarAlgebra.toNonUnitalNormedCommRing
+    (A : Type*) [NonUnitalCommCStarAlgebra A] : NonUnitalNormedCommRing A where
+  mul_comm := mul_comm
+
+/-- This is not registered as an instance to avoid Lean searching for `IsClosed (s : Set A)`
+instances frequently. -/
```
Comment (eric-wieser) [inline Mathlib/Analysis/CStarAlgebra/Classes.lean:55]: I feel like this is the whole point of `IsClosed` being a typeclass though...

File: Mathlib/Analysis/CStarAlgebra/Classes.lean [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+instance [(i : ι) → NonUnitalCommCStarAlgebra (A i)] : NonUnitalCommCStarAlgebra (Π i, A i) where
+  mul_comm := mul_comm
+
+noncomputable instance [(i : ι) → CStarAlgebra (A i)] : CStarAlgebra (Π i, A i) where
+
+noncomputable instance [(i : ι) → CommCStarAlgebra (A i)] : CommCStarAlgebra (Π i, A i) where
+  mul_comm := mul_comm
+
+end Pi
```
Comment (eric-wieser) [inline Mathlib/Analysis/CStarAlgebra/Classes.lean:109]: `ULift` and `PUnit` would be good instances too, assuming the machinery is present elsewhere.

File: Mathlib/Topology/ContinuousFunction/Algebra.lean [hunk 5]
```diff
@@ -579,6 +579,14 @@ instance [SMul R M] [ContinuousConstSMul R M] [SMul R₁ M] [ContinuousConstSMul
 instance [SMul R M] [SMul Rᵐᵒᵖ M] [ContinuousConstSMul R M] [IsCentralScalar R M] :
     IsCentralScalar R C(α, M) where op_smul_eq_smul _ _ := ext fun _ => op_smul_eq_smul _ _
 
+instance [SMul R M] [ContinuousConstSMul R M] [Mul M] [ContinuousMul M] [IsScalarTower R M M] :
+    IsScalarTower R C(α, M) C(α, M) where
+  smul_assoc _ _ _ := ext fun _ => smul_mul_assoc ..
+
+instance [SMul R M] [ContinuousConstSMul R M] [Mul M] [ContinuousMul M] [SMulCommClass R M M] :
+    SMulCommClass R C(α, M) C(α, M) where
+  smul_comm _ _ _ := ext fun _ => (mul_smul_comm ..).symm
```
Comment (eric-wieser) [inline Mathlib/Topology/ContinuousFunction/Algebra.lean:588]: These instances and the ones elsewhere in this PR might make sense to spin out, in order to make this PR a bit smaller and to allow more scrutiny (in particular, you're missing the symmetric version of the `SMulCommClass` one).

File: Mathlib/Analysis/CStarAlgebra/Unitization.lean [hunk 6]
```diff
@@ -171,4 +171,11 @@ instance Unitization.instCStarRing : CStarRing (Unitization 𝕜 E) where
     · replace h := (not_le.mp h).le
       rw [sq, sq, sup_eq_left.mpr h, sup_eq_left.mpr (mul_self_le_mul_self (norm_nonneg _) h)]
 
+noncomputable instance Unitization.instCStarAlgebra {A : Type*} [CStarAlgebra A] :
+    CStarAlgebra (Unitization ℂ A) where
```
Comment (dupuisf) [inline Mathlib/Analysis/CStarAlgebra/Unitization.lean:175]: Shouldn't we just require `NonUnitalCStarAlgebra A` here?

**Comment by j-loreaux** [comment]
Closing in favor of #16953 to avoid a performance hit.

**Pull request closed without merging**